src/metrics/rules/active_ruleset.go

```diff
@@ -45,6 +48,69 @@ type activeRuleSet struct {
 	tagsFilterOpts  filters.TagsFilterOptions
 	newRollupIDFn   metricid.NewIDFn
 	isRollupIDFn    metricid.MatchIDFn
+
+	useFastMatch bool
+	// The key is the value of the __name__ label, the value is the mappingRule
+	mappingRuleIndex map[string][]*mappingRule
+	// The key is the value of the __name__ label, the value is rollupRule
+	rollupRuleIndex map[string][]*rollupRule
+}
+
+func newActiveRuleSetFast(
+	version int,
+	mappingRules []*mappingRule,
+	rollupRules []*rollupRule,
+	tagsFilterOpts filters.TagsFilterOptions,
+	newRollupIDFn metricid.NewIDFn,
+	isRollupIDFn metricid.MatchIDFn,
+) *activeRuleSet {
+	rset := newActiveRuleSet(version, mappingRules, rollupRules, tagsFilterOpts, newRollupIDFn, isRollupIDFn)
+	rset.useFastMatch = true
+
+	nowNanos := time.Now().UnixNano()
+
+	getNameTagValue := func(nameFilterValue *filters.FilterValue, snapshotString func() string) string {
+		if nameFilterValue == nil {
+			panic(fmt.Sprintf("A rule has no fitler for metric name tag: %s", snapshotString()))
+		}
+		if nameFilterValue.Negate {
+			panic(fmt.Sprintf("A rule has negated filter for metric name tag: %s", snapshotString()))
+		}
+		if strings.Contains(nameFilterValue.Pattern, "*") {
+			panic(fmt.Sprintf("A rule has wildcard filter for metric name tag %s", snapshotString()))
+		}
+		return nameFilterValue.Pattern
+	}
+
+	rset.mappingRuleIndex = make(map[string][]*mappingRule)
+	for _, rule := range mappingRules {
+		snapshot := rule.activeSnapshot(nowNanos)
+		if snapshot == nil {
+			panic(fmt.Sprintf("mapping rule has no active snapshot: %+v", rule))
+		}
+		nameTagValue := getNameTagValue(snapshot.filter.NameFilterValue(), func() string { return fmt.Sprintf("%+v", snapshot) })
+		rset.mappingRuleIndex[nameTagValue] = append(rset.mappingRuleIndex[nameTagValue], rule)
+	}
+	rset.rollupRuleIndex = make(map[string][]*rollupRule)
```

Comment:
Not a big thing, IIRC rollupRule has to come with a mapping rule first, maybe mapping rule is enough to decide to skip or not.

Reply:
Right, we enforce that in the jsonnet util function. I'd prefer to be defensive in the Golang code.